src/discordtipbot/cache.cr

```diff
@@ -0,0 +1,58 @@
+class Cache
+  # Container for a snowflake and the last time it was seen
+  record Entry, id : UInt64, seen : Time do
+    def ==(other)
+      id == other.id
+    end
+
+    def hash
+      id.hash
+    end
+
+    def older_than?(age : Time::Span)
+      Time.now > seen + age
+    end
+  end
+
+  @cache = {} of UInt64 => Set(Entry)
+
+  def initialize(@ttl : Time::Span)
+  end
+
+  # Marks the `user_id` as being observed in `channel_id` at the given `time`
+  def touch(channel_id : UInt64, user_id : UInt64, time = Time.now)
+    entry = Entry.new(user_id, time)
```

Comment:
Should we also touch for users where ttl has already passed?

Reply:
When/why would you do that?

Reply:
When parsing old messages of an channel